### PR TITLE
add mock providers for testing framework

### LINKED ISSUE #1155: Support for mocking providers in opentofu test
### OpenTofu Version

```shell
OpenTofu v1.6.0
```

### Use Cases

Test OpenTofu code offline.

### Attempted Solutions

Using fake credentials in your provider in the testfile does not always work, the newly implemented _mock_ provider by Terraform has a broader usecase, as it allows faking data sources etc.

https://developer.hashicorp.com/terraform/language/tests/mocking

For example, using fake credentials and trying to retrieve an AMI with

```
data "aws_ami" "ubuntu" {
  most_recent = true

  filter {
    name   = "name"
    values = ["ubuntu/images/hvm-ssd/ubuntu-jammy-22.04-amd64-server-*"]
  }

  filter {
    name   = "virtualization-type"
    values = ["hvm"]
  }

  owners = ["099720109477"] # Canonical
}
```

Returns the following error

```
reading EC2 AMIs: AuthFailure: AWS was not able to validate the provided access credentials
```

### Proposal

Add support for mocking providers, resources, and data sources.

### References

_No response_

### PR DESCRIPTION
**Pull request by ollevche**
Resolves #1155.

This PR adds support for `mock_provider` block. It reuses a mock value generation from the previous override implementation. As a part of this PR, I changed mock values composition to be stable (i.e. produce the same result for the same input) so `providerForTest` doesn't need to cache planned values.

TODO:
- [X] extend existing e2e test to cover `mock_provider`
- [X] extend test command website documentation
- [ ] ~~`.tfmock.hcl` / `.tofumock.hcl` special extensions for mock data~~ (moved to a separate issue https://github.com/opentofu/opentofu/issues/1778)

## Target Release

1.8.0

## Checklist

- [X] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [X] I have not used an AI coding assistant to create this PR.
- [X] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.

### Go checklist

 

- [X] I have run golangci-lint on my change and receive no errors relevant to my code.
- [X] I have run existing tests to ensure my code doesn't break anything.
- [X] I have added tests for all relevant use cases of my code, and those tests are passing.
- [X] I have only exported functions, variables and structs that should be used from other packages.
- [X] I have added meaningful comments to all exported functions, variables, and structs.

### CODE REVIEW

**Comment by ollevche** [comment]
TODO: format tofu test code in a separate PR

File: internal/configs/hcl2shim/mock_value_composer.go [hunk 1]
```diff
@@ -310,7 +312,9 @@ func (mvc mockValueComposer) getMockValueByType(t cty.Type) (cty.Value, bool) {
 		objVals := make(map[string]cty.Value)
 
 		// populate the object with mock values
-		for k, at := range t.AttributeTypes() {
+		for _, kv := range mapToSortedSlice(t.AttributeTypes()) {
```
Comment (Yantrio) [inline internal/configs/hcl2shim/mock_value_composer.go:315]: Can you add a comment to explain why we want to iterate over the attribute types in a sorted manner please?

**Review by Yantrio — APPROVED** [review]
Nice work! Just one tiny nitpick about sorted map iteration

**Review by Yantrio — DISMISSED** [review]

File: internal/tofu/eval_context_builtin.go [hunk 2]
```diff
@@ -145,6 +147,18 @@ func (ctx *BuiltinEvalContext) InitProvider(addr addrs.AbsProviderConfig) (provi
 		return nil, err
 	}
 
+	if ctx.Config != nil {
```
Comment (cam72cam) [inline internal/tofu/eval_context_builtin.go:150]: The config is already accessed in EvaluationScope below:
https://github.com/opentofu/opentofu/pull/1772/files?diff=split&w=1#diff-126d6884fd43891cbb56845ee6f7f51ed02895af69190815e5c694a51a3777d3L522

I don't think we need to add ctx.Config and can use that same code instead.

Comment (ollevche) [reply internal/tofu/eval_context_builtin.go]: Nice catch, thanks!

[5fa601a](https://github.com/opentofu/opentofu/pull/1772/commits/5fa601a8bae2f435f88d08e2b390d95febabbb6d)

**Review by cam72cam — APPROVED** [review]

**Pull request merged**